File: backend/controllers/authController.js
import { getFirestore, admin } from '../config/firebase-config.js';
import dotenv from 'dotenv';

dotenv.config();

// Verify Firebase ID token middleware
export const verifyToken = async (req, res, next) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];
    
    if (!token) {
      return res.status(401).json({ error: 'No token provided' });
    }

    const decodedToken = await admin.auth().verifyIdToken(token);
    req.user = decodedToken;
    next();
  } catch (error) {
    console.error('Token verification error:', error);
    res.status(401).json({ error: 'Invalid token' });
  }
};

// Get current user profile
export const getUserProfile = async (req, res) => {
  try {
    const { uid } = req.user;
    const db = getFirestore();
    
    // Get user from Firestore
    const userDoc = await db.collection('users').doc(uid).get();
    
    // If user doesn't exist in Firestore, create from Firebase Auth
    if (!userDoc.exists) {
      const firebaseUser = await admin.auth().getUser(uid);
      const userData = {
        uid: firebaseUser.uid,
        email: firebaseUser.email,
        displayName: firebaseUser.displayName || null,
        photoURL: firebaseUser.photoURL || null,
        emailVerified: firebaseUser.emailVerified,
        createdAt: new Date(firebaseUser.metadata.creationTime),
      };
      
      await db.collection('users').doc(uid).set(userData);
      
      return res.json({
        success: true,
        user: userData
      });
    }
    
    const userData = userDoc.data();
    res.json({
      success: true,
      user: {
        uid: userData.uid,
        email: userData.email,
        displayName: userData.displayName,
        photoURL: userData.photoURL,
        emailVerified: userData.emailVerified,
        createdAt: userData.createdAt,
      }
    });
  } catch (error) {
    console.error('Get profile error:', error);
    res.status(500).json({ error: 'Failed to get user profile' });
  }
};

// Update user profile
export const updateUserProfile = async (req, res) => {
  try {
    const { uid } = req.user;
    const updates = req.body;
    const db = getFirestore();
    
    // Remove sensitive fields
    delete updates.uid;
    delete updates.email;
    delete updates.emailVerified;
    
    const userRef = db.collection('users').doc(uid);
    const userDoc = await userRef.get();
    
    if (!userDoc.exists) {
      return res.status(404).json({ error: 'User not found' });
    }
    
    await userRef.update({
      ...updates,
      updatedAt: new Date()
    });
    
    const updatedUser = await userRef.get();
    const userData = updatedUser.data();
    
    res.json({
      success: true,
      message: 'Profile updated successfully',
      user: {
        uid: userData.uid,
        email: userData.email,
        displayName: userData.displayName,
        photoURL: userData.photoURL,
        emailVerified: userData.emailVerified,
      }
    });
  } catch (error) {
    console.error('Update profile error:', error);
    res.status(500).json({ error: 'Failed to update profile' });
  }
};

// Set custom claims for user
export const setCustomClaims = async (req, res) => {
  try {
    const { uid } = req.params;
    const { claims } = req.body;
    const db = getFirestore();
    
    await admin.auth().setCustomUserClaims(uid, claims);
    
    // Update in Firestore
    const userRef = db.collection('users').doc(uid);
    const userDoc = await userRef.get();
    
    if (userDoc.exists) {
      await userRef.update({ 
        customClaims: claims,
        updatedAt: new Date()
      });
    }
    
    res.json({
      success: true,
      message: 'Custom claims set successfully'
    });
  } catch (error) {
    console.error('Set custom claims error:', error);
    res.status(500).json({ error: 'Failed to set custom claims' });
  }
};

// Delete user
export const deleteUser = async (req, res) => {
  try {
    const { uid } = req.params;
    const db = getFirestore();
    
    // Delete from Firebase Auth
    await admin.auth().deleteUser(uid);
    
    // Delete from Firestore
    await db.collection('users').doc(uid).delete();
    
    res.json({
      success: true,
      message: 'User deleted successfully'
    });
  } catch (error) {
    console.error('Delete user error:', error);
    res.status(500).json({ error: 'Failed to delete user' });
  }
};

export const registerUser = async (req, res) => {
  try {
    const { email, password } = req.body;

    const response = await fetch(
      `https://identitytoolkit.googleapis.com/v1/accounts:signUp?key=${process.env.FIREBASE_API_KEY}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          email,
          password,
          returnSecureToken: true,
        }),
      }
    );

    const user = await response.json();

    if (!response.ok) {
      return res.status(400).json({ 
        success: false, 
        message: user.error?.message || 'Registration failed' 
      });
    }

    // Save user to Firestore
    const db = getFirestore();
    await db.collection('users').doc(user.localId).set({
      uid: user.localId,
      email,
      emailVerified: false,
      createdAt: new Date(),
    });

    return res.status(200).json({ 
      success: true, 
      message: 'User registered successfully',
      id: user.localId 
    });

  } catch (err) {
    return res.status(500).json({ 
      success: false, 
      message: err.message 
    });
  }
};

export const loginUser = async (req, res) => {
  try {
    const { email, password } = req.body;

    const response = await fetch(
      `https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${process.env.FIREBASE_API_KEY}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ email, password, returnSecureToken: true }),
      }
    );

    const data = await response.json();

    if (!response.ok) {
      return res.status(400).json({ 
        success: false, 
        message: data.error?.message || 'Login failed' 
      });
    }

    return res.json({ 
      success: true, 
      message: 'Login successful', 
      token: data.idToken 
    });
  } catch (err) {
    return res.status(500).json({ 
      success: false, 
      message: err.message 
    });
  }
};

File: functions/controllers/auth_controller.py
from firebase_admin import auth, firestore
from flask import request, jsonify
import functools
from datetime import datetime

def get_db():
    """Get Firestore client instance"""
    return firestore.client()

def verify_token(f):
    @functools.wraps(f)
    def decorated_function(*args, **kwargs):
        try:
            auth_header = request.headers.get('Authorization')
            if not auth_header or not auth_header.startswith('Bearer '):
                return jsonify({'error': 'No token provided'}), 401
            
            token = auth_header.split(' ')[1]
            decoded_token = auth.verify_id_token(token)
            request.user = decoded_token
            return f(*args, **kwargs)
        except Exception as e:
            print(f'Token verification error: {e}')
            return jsonify({'error': 'Invalid token'}), 401
    
    return decorated_function

def create_user_profile(uid, email, display_name=None, additional_data=None):
    """Create user profile in Firestore using enhanced User model"""
    try:
        from models.user import User
        
        # Create User object with enhanced data
        user = User(
            uid=uid,
            email=email,
            display_name=display_name or email.split('@')[0],
            phone=additional_data.get('phone') if additional_data else None,
            bio=additional_data.get('bio') if additional_data else None,
            profile_complete=additional_data.get('profile_complete', False) if additional_data else False
        )
        
        # Validate and save
        validation_errors = user.validate()
        if validation_errors:
            print(f'User validation errors: {validation_errors}')
            return None
        
        success = user.save()
        if success:
            return user.to_dict()
        else:
            return None
            
    except Exception as e:
        print(f'Error creating user profile: {e}')
        return None

def get_user_profile():
    """Get user profile from Firestore using enhanced User model"""
    try:
        uid = request.user['uid']
        from models.user import User
        
        # Try to get existing user
        user = User.get_by_id(uid)
        
        if user:
            return user.to_dict()
        else:
            # Create profile if doesn't exist (for existing Firebase Auth users)
            user_record = auth.get_user(uid)
            return create_user_profile(uid, user_record.email, user_record.display_name)
    except Exception as e:
        print(f'Error getting user profile: {e}')
        return None

def update_user_profile(uid, update_data):
    """Update user profile in Firestore using enhanced User model"""
    try:
        from models.user import User
        
        user = User.get_by_id(uid)
        if user:
            return user.update(update_data)
        else:
            print(f'User not found: {uid}')
            return False
    except Exception as e:
        print(f'Error updating user profile: {e}')
        return False

def register_user():
    try:
        data = request.get_json()
        email = data.get('email')
        password = data.get('password')
        
        user_record = auth.create_user(
            email=email,
            password=password,
            email_verified=False
        )

        create_user_profile(user_record.uid, email)

        return jsonify({
            'success': True,
            'message': 'User registered successfully',
            'uid': user_record.uid
        }), 201

    except Exception as e:
        print(f'Registration error: {e}')
        return jsonify({
            'success': False,
            'message': str(e)
        }), 400

def login_user():
    try:
        return jsonify({
            'success': True,
            'message': 'Use Firebase Auth SDK for authentication'
        })
    except Exception as e:
        print(f'Login error: {e}')
        return jsonify({'error': 'Login failed'}), 500